File: Algorithms/576. Out of Boundary Paths.rs
const div: i32 = 1_000_000_007;

impl Solution {
    pub fn find_paths(m: i32, n: i32, max_move: i32, start_row: i32, start_column: i32) -> i32 {
        let rows = m as usize;
        let columns = n as usize;
        let mut dp = vec![vec![0; columns]; rows];
        dp[start_row as usize][start_column as usize] = 1;
        let mut res = 0;
        for moves in 1..=max_move {
            let mut temp = vec![vec![0; columns]; rows];
            for i in 0..rows {
                for j in 0..columns {
                    if i == rows - 1 { res = (res + dp[i][j]) % div; }
                    if j == columns - 1 { res = (res + dp[i][j]) % div; }
                    if i == 0 { res = (res + dp[i][j]) % div; }
                    if j == 0 { res = (res + dp[i][j]) % div; }
                    let mut temp_num_1 = 0;
                    if i > 0 { temp_num_1 += dp[i - 1][j]; }
                    if i < rows - 1 { temp_num_1 +=  dp[i + 1][j]; }
                    temp_num_1 %= div;
                    let mut temp_num_2 = 0;
                    if j > 0 { temp_num_2 += dp[i][j - 1]; }
                    if j < columns - 1  { temp_num_2 +=  dp[i][j + 1]; }
                    temp_num_2 %= div;

                    temp[i][j] = (temp_num_1 + temp_num_2) % div;
                }
            }
            dp = temp;
        }
        res
    }
}

File: Algorithms/576. Out of Boundary Paths.ts
function findPaths(
  m: number,
  n: number,
  maxMove: number,
  startRow: number,
  startColumn: number
): number {
  const MOD = 1_000_000_007;
  let dp = Array.from({ length: m }, (_) => new Array(n).fill(0));
  dp[startRow][startColumn] = 1;
  let result = 0;

  for (let moves = 1; moves <= maxMove; moves++) {
    let temp = Array.from({ length: m }, (_) => new Array(n).fill(0));
    for (let i = 0; i < m; i++) {
      for (let j = 0; j < n; j++) {
        if (i === m - 1) {
          result = (result + dp[i][j]) % MOD;
        }
        if (j === n - 1) {
          result = (result + dp[i][j]) % MOD;
        }
        if (i === 0) {
          result = (result + dp[i][j]) % MOD;
        }
        if (j === 0) {
          result = (result + dp[i][j]) % MOD;
        }
        temp[i][j] =
          ((((i > 0 ? dp[i - 1][j] : 0) + (i < m - 1 ? dp[i + 1][j] : 0)) %
            MOD) +
            (((j > 0 ? dp[i][j - 1] : 0) + (j < n - 1 ? dp[i][j + 1] : 0)) %
              MOD)) %
          MOD;
      }
    }
    dp = temp;
  }

  return result;
}
